build restatectl failed without features `storage-query`

At commit `93fd7913a4f19501557cefb97b72b03188a7010d`, run command `cargo build -p restatectl` failed with:

```
error[E0433]: failed to resolve: use of undeclared crate or module `datafusion`
  --> crates/admin/src/cluster_controller/grpc_svc_handler.rs:12:5
   |
12 | use datafusion::arrow::ipc::writer::StreamWriter;
   |     ^^^^^^^^^^ use of undeclared crate or module `datafusion`

error[E0433]: failed to resolve: use of undeclared crate or module `datafusion`
  --> crates/admin/src/cluster_controller/grpc_svc_handler.rs:13:5
   |
13 | use datafusion::error::DataFusionError;
   |     ^^^^^^^^^^ use of undeclared crate or module `datafusion`

error[E0433]: failed to resolve: use of undeclared crate or module `restate_storage_query_datafusion`
  --> crates/admin/src/cluster_controller/grpc_svc_handler.rs:22:5
   |
22 | use restate_storage_query_datafusion::context::QueryContext;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `restate_storage_query_datafusion`

error[E0432]: unresolved import `crate::storage_query`
  --> crates/admin/src/cluster_controller/grpc_svc_handler.rs:40:12
   |
40 | use crate::storage_query::WriteRecordBatchStream;
   |            ^^^^^^^^^^^^^ could not find `storage_query` in the crate root
   |
note: found an item that was configured out
  --> crates/admin/src/lib.rs:20:5
   |
20 | mod storage_query;
   |     ^^^^^^^^^^^^^
note: the item is gated behind the `storage-query` feature
  --> crates/admin/src/lib.rs:19:7
   |
19 | #[cfg(feature = "storage-query")]
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0433]: failed to resolve: use of undeclared crate or module `restate_storage_query_datafusion`
  --> crates/admin/src/cluster_controller/service.rs:20:5
   |
20 | use restate_storage_query_datafusion::context::{ClusterTables, QueryContext};
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `restate_storage_query_datafusion`

error[E0432]: unresolved import `restate_storage_query_datafusion`
  --> crates/admin/src/cluster_controller/service.rs:19:5
   |
19 | use restate_storage_query_datafusion::BuildError;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `restate_storage_query_datafusion`

^C  Building [=====================> ] 1112/1115: restate-admin
```

if we change command to `cargo build -p restatectl --features "restate-admin/storage-query"`, it works as expect.

--- a/crates/admin/Cargo.toml
+++ b/crates/admin/Cargo.toml
@@ -14,7 +14,7 @@ options_schema = ["restate-service-client/options_schema"]
 memory-loglet = ["restate-bifrost/memory-loglet"]
 replicated-loglet = ["restate-bifrost/replicated-loglet"]
 serve-web-ui = ["restate-web-ui", "mime_guess"]
-storage-query = ["dep:restate-storage-query-datafusion", "dep:datafusion"]
+storage-query = []
 metadata-api = []
 
 [dependencies]
@@ -29,7 +29,7 @@ restate-futures-util = { workspace = true }
 restate-metadata-server = { workspace = true }
 restate-service-client = { workspace = true }
 restate-service-protocol = { workspace = true, features = ["discovery"] }
-restate-storage-query-datafusion = { workspace = true, optional = true }
+restate-storage-query-datafusion = { workspace = true }
 restate-types = { workspace = true, features = ["schemars"] }
 restate-utoipa = { workspace = true }
 restate-wal-protocol = { workspace = true }
@@ -41,7 +41,7 @@ axum = { workspace = true, features = ["json"] }
 bytes = { workspace = true }
 bytestring = { workspace = true }
 codederror = { workspace = true }
-datafusion = { workspace = true, optional = true }
+datafusion = { workspace = true }
 derive_builder = { workspace = true }
 derive_more = { workspace = true }
 enumset = { workspace = true }

--- a/crates/admin/src/cluster_controller/grpc_svc_handler.rs
+++ b/crates/admin/src/cluster_controller/grpc_svc_handler.rs
@@ -37,7 +37,7 @@ use crate::cluster_controller::protobuf::{
     FindTailResponse, ListLogsRequest, ListLogsResponse, SealAndExtendChainRequest,
     SealAndExtendChainResponse, SealedSegment, TailState, TrimLogRequest,
 };
-use crate::storage_query::WriteRecordBatchStream;
+use crate::query_utils::WriteRecordBatchStream;
 
 use super::ClusterControllerHandle;
 use super::protobuf::{

--- a/crates/admin/src/lib.rs
+++ b/crates/admin/src/lib.rs
@@ -12,6 +12,7 @@ pub mod cluster_controller;
 mod error;
 #[cfg(feature = "metadata-api")]
 mod metadata_api;
+mod query_utils;
 mod rest_api;
 mod schema_registry;
 pub mod service;

--- a/crates/admin/src/query_utils.rs
+++ b/crates/admin/src/query_utils.rs
@@ -1,0 +1,174 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+use std::io::Write;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use bytes::Bytes;
+use datafusion::arrow::datatypes::Schema;
+use datafusion::arrow::ipc::writer::StreamWriter;
+use datafusion::arrow::json::writer::JsonArray;
+use datafusion::arrow::record_batch::RecordBatch;
+use datafusion::error::DataFusionError;
+use datafusion::execution::SendableRecordBatchStream;
+use futures::{Stream, StreamExt, ready};
+use parking_lot::Mutex;
+use tracing::{Level, enabled, warn};
+
+pub trait RecordBatchWriter
+where
+    Self: Sized,
+{
+    // Create the writer
+    fn new(schema: &Schema) -> Result<Self, DataFusionError>;
+
+    /// Write a single batch to the writer.
+    fn write(&mut self, batch: &RecordBatch) -> Result<Bytes, DataFusionError>;
+
+    /// Write footer or termination data, then mark the writer as done.
+    fn finish(&mut self) -> Result<Bytes, DataFusionError>;
+}
+
+impl RecordBatchWriter for StreamWriter<Vec<u8>> {
+    fn new(schema: &Schema) -> Result<Self, DataFusionError> {
+        Ok(Self::try_new(Vec::new(), schema)?)
+    }
+
+    fn write(&mut self, batch: &RecordBatch) -> Result<Bytes, DataFusionError> {
+        self.write(batch)?;
+        let bytes = Bytes::copy_from_slice(self.get_ref());
+        self.get_mut().clear();
+        Ok(bytes)
+    }
+
+    fn finish(&mut self) -> Result<Bytes, DataFusionError> {
+        self.finish()?;
+        let bytes = Bytes::copy_from_slice(self.get_ref());
+        self.get_mut().clear();
+        Ok(bytes)
+    }
+}
+
+#[derive(Clone)]
+// unfortunately the json writer doesnt give a way to get a mutable reference to the underlying writer, so we need another pointer in to its buffer
+// we use a lock here to help make the writer send/sync, despite it being totally uncontended :(
+struct LockWriter(Arc<Mutex<Vec<u8>>>);
+
+impl LockWriter {
+    fn new() -> Self {
+        Self(Arc::new(Mutex::new(Vec::new())))
+    }
+
+    fn take(&self) -> Vec<u8> {
+        let mut vec = self.0.lock();
+        let new_vec = Vec::with_capacity(vec.capacity());
+        std::mem::replace(&mut vec, new_vec)
+    }
+}
+
+impl Write for LockWriter {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.0.lock().write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.lock().flush()
+    }
+}
+
+pub struct JsonWriter {
+    json_writer: datafusion::arrow::json::Writer<LockWriter, JsonArray>,
+    lock_writer: LockWriter,
+    finished: bool,
+}
+
+impl RecordBatchWriter for JsonWriter {
+    fn new(_schema: &Schema) -> Result<Self, DataFusionError> {
+        let mut lock_writer = LockWriter::new();
+        // we write out under 'rows' key so that we may add extra keys later (eg 'schema')
+        lock_writer.write_all(br#"{"rows":"#)?;
+        Ok(Self {
+            json_writer: datafusion::arrow::json::Writer::new(lock_writer.clone()),
+            lock_writer,
+            finished: false,
+        })
+    }
+
+    fn write(&mut self, batch: &RecordBatch) -> Result<Bytes, DataFusionError> {
+        self.json_writer.write(batch)?;
+        Ok(Bytes::from(self.lock_writer.take()))
+    }
+
+    fn finish(&mut self) -> Result<Bytes, DataFusionError> {
+        if !self.finished {
+            self.finished = true;
+
+            self.json_writer.finish()?;
+            self.lock_writer.write_all(b"}")?;
+        }
+        Ok(Bytes::from(self.lock_writer.take()))
+    }
+}
+
+pub struct WriteRecordBatchStream<W> {
+    done: bool,
+    record_batch_stream: SendableRecordBatchStream,
+    stream_writer: W,
+    query: String,
+}
+
+impl<W: RecordBatchWriter> WriteRecordBatchStream<W> {
+    pub fn new(
+        record_batch_stream: SendableRecordBatchStream,
+        query: String,
+    ) -> Result<Self, DataFusionError> {
+        Ok(WriteRecordBatchStream {
+            done: false,
+            stream_writer: W::new(&record_batch_stream.schema())?,
+            record_batch_stream,
+            query,
+        })
+    }
+}
+
+impl<W: RecordBatchWriter + Unpin> Stream for WriteRecordBatchStream<W> {
+    type Item = Result<Bytes, DataFusionError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.done {
+            return Poll::Ready(None);
+        }
+
+        let record_batch = ready!(self.record_batch_stream.poll_next_unpin(cx));
+
+        if let Some(record_batch) = record_batch {
+            match record_batch.and_then(|record_batch| self.stream_writer.write(&record_batch)) {
+                Ok(bytes) => Poll::Ready(Some(Ok(bytes))),
+                Err(err) => {
+                    if enabled!(Level::DEBUG) {
+                        warn!(query = %self.query, %err, "Query failed");
+                    } else {
+                        warn!(%err, "Query failed");
+                    }
+
+                    self.done = true;
+                    Poll::Ready(Some(Err(err)))
+                }
+            }
+        } else {
+            self.done = true;
+            match self.stream_writer.finish() {
+                Err(err) => Poll::Ready(Some(Err(err))),
+                Ok(bytes) => Poll::Ready(Some(Ok(bytes))),
+            }
+        }
+    }
+}

--- a/crates/admin/src/storage_query/mod.rs
+++ b/crates/admin/src/storage_query/mod.rs
@@ -15,7 +15,6 @@ mod query;
 use axum::{Router, routing::post};
 use std::sync::Arc;
 
-pub use query::{RecordBatchWriter, WriteRecordBatchStream};
 use restate_storage_query_datafusion::context::QueryContext;
 
 #[derive(Clone)]

--- a/crates/admin/src/storage_query/query.rs
+++ b/crates/admin/src/storage_query/query.rs
@@ -8,10 +8,9 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use std::io::Write;
-use std::pin::Pin;
 use std::sync::Arc;
-use std::task::{Context, Poll};
+
+use crate::query_utils::{JsonWriter, WriteRecordBatchStream};
 
 use super::QueryServiceState;
 use super::convert::{ConvertRecordBatchStream, V1_CONVERTER};
@@ -19,24 +18,17 @@ use super::error::StorageQueryError;
 use axum::extract::State;
 use axum::response::{IntoResponse, Response};
 use axum::{Extension, Json, http};
-use bytes::Bytes;
-use datafusion::arrow::datatypes::Schema;
 use datafusion::arrow::ipc::writer::StreamWriter;
-use datafusion::arrow::json::writer::JsonArray;
-use datafusion::arrow::record_batch::RecordBatch;
-use datafusion::error::DataFusionError;
 use datafusion::execution::SendableRecordBatchStream;
-use futures::{Stream, StreamExt, TryStreamExt, ready};
+use futures::{StreamExt, TryStreamExt};
 use http::{HeaderMap, HeaderValue};
 use http_body::Frame;
 use http_body_util::StreamBody;
 use okapi_operation::*;
-use parking_lot::Mutex;
 use restate_admin_rest_model::version::AdminApiVersion;
 use schemars::JsonSchema;
 use serde::Deserialize;
 use serde_with::serde_as;
-use tracing::{Level, enabled, warn};
 
 #[serde_as]
 #[derive(Debug, Deserialize, JsonSchema)]
@@ -96,154 +88,4 @@ pub async fn query(
         .header(http::header::CONTENT_TYPE, content_type)
         .body(StreamBody::new(result_stream))
         .expect("content-type header is correct"))
-}
-
-pub trait RecordBatchWriter
-where
-    Self: Sized,
-{
-    // Create the writer
-    fn new(schema: &Schema) -> Result<Self, DataFusionError>;
-
-    /// Write a single batch to the writer.
-    fn write(&mut self, batch: &RecordBatch) -> Result<Bytes, DataFusionError>;
-
-    /// Write footer or termination data, then mark the writer as done.
-    fn finish(&mut self) -> Result<Bytes, DataFusionError>;
-}
-
-impl RecordBatchWriter for StreamWriter<Vec<u8>> {
-    fn new(schema: &Schema) -> Result<Self, DataFusionError> {
-        Ok(Self::try_new(Vec::new(), schema)?)
-    }
-
-    fn write(&mut self, batch: &RecordBatch) -> Result<Bytes, DataFusionError> {
-        self.write(batch)?;
-        let bytes = Bytes::copy_from_slice(self.get_ref());
-        self.get_mut().clear();
-        Ok(bytes)
-    }
-
-    fn finish(&mut self) -> Result<Bytes, DataFusionError> {
-        self.finish()?;
-        let bytes = Bytes::copy_from_slice(self.get_ref());
-        self.get_mut().clear();
-        Ok(bytes)
-    }
-}
-
-#[derive(Clone)]
-// unfortunately the json writer doesnt give a way to get a mutable reference to the underlying writer, so we need another pointer in to its buffer
-// we use a lock here to help make the writer send/sync, despite it being totally uncontended :(
-struct LockWriter(Arc<Mutex<Vec<u8>>>);
-
-impl LockWriter {
-    fn new() -> Self {
-        Self(Arc::new(Mutex::new(Vec::new())))
-    }
-
-    fn take(&self) -> Vec<u8> {
-        let mut vec = self.0.lock();
-        let new_vec = Vec::with_capacity(vec.capacity());
-        std::mem::replace(&mut vec, new_vec)
-    }
-}
-
-impl Write for LockWriter {
-    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
-        self.0.lock().write(buf)
-    }
-
-    fn flush(&mut self) -> std::io::Result<()> {
-        self.0.lock().flush()
-    }
-}
-
-struct JsonWriter {
-    json_writer: datafusion::arrow::json::Writer<LockWriter, JsonArray>,
-    lock_writer: LockWriter,
-    finished: bool,
-}
-
-impl RecordBatchWriter for JsonWriter {
-    fn new(_schema: &Schema) -> Result<Self, DataFusionError> {
-        let mut lock_writer = LockWriter::new();
-        // we write out under 'rows' key so that we may add extra keys later (eg 'schema')
-        lock_writer.write_all(br#"{"rows":"#)?;
-        Ok(Self {
-            json_writer: datafusion::arrow::json::Writer::new(lock_writer.clone()),
-            lock_writer,
-            finished: false,
-        })
-    }
-
-    fn write(&mut self, batch: &RecordBatch) -> Result<Bytes, DataFusionError> {
-        self.json_writer.write(batch)?;
-        Ok(Bytes::from(self.lock_writer.take()))
-    }
-
-    fn finish(&mut self) -> Result<Bytes, DataFusionError> {
-        if !self.finished {
-            self.finished = true;
-
-            self.json_writer.finish()?;
-            self.lock_writer.write_all(b"}")?;
-        }
-        Ok(Bytes::from(self.lock_writer.take()))
-    }
-}
-
-pub struct WriteRecordBatchStream<W> {
-    done: bool,
-    record_batch_stream: SendableRecordBatchStream,
-    stream_writer: W,
-    query: String,
-}
-
-impl<W: RecordBatchWriter> WriteRecordBatchStream<W> {
-    pub fn new(
-        record_batch_stream: SendableRecordBatchStream,
-        query: String,
-    ) -> Result<Self, DataFusionError> {
-        Ok(WriteRecordBatchStream {
-            done: false,
-            stream_writer: W::new(&record_batch_stream.schema())?,
-            record_batch_stream,
-            query,
-        })
-    }
-}
-
-impl<W: RecordBatchWriter + Unpin> Stream for WriteRecordBatchStream<W> {
-    type Item = Result<Bytes, DataFusionError>;
-
-    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
-        if self.done {
-            return Poll::Ready(None);
-        }
-
-        let record_batch = ready!(self.record_batch_stream.poll_next_unpin(cx));
-
-        if let Some(record_batch) = record_batch {
-            match record_batch.and_then(|record_batch| self.stream_writer.write(&record_batch)) {
-                Ok(bytes) => Poll::Ready(Some(Ok(bytes))),
-                Err(err) => {
-                    if enabled!(Level::DEBUG) {
-                        warn!(query = %self.query, %err, "Query failed");
-                    } else {
-                        warn!(%err, "Query failed");
-                    }
-
-                    self.done = true;
-                    Poll::Ready(Some(Err(err)))
-                }
-            }
-        } else {
-            self.done = true;
-            match self.stream_writer.finish() {
-                Err(err) => Poll::Ready(Some(Err(err))),
-                Ok(bytes) => Poll::Ready(Some(Ok(bytes))),
-            }
-        }
-    }
 }

